sbi/analysis/sbc.py

```diff
@@ -0,0 +1,306 @@
+        with tqdm_joblib(
+            tqdm(
+                thos_batches,
+                disable=not show_progress_bar,
+                desc=f"""Running {num_sbc_samples} sbc runs in {len(thos_batches)}
+                    batches.""",
+                total=len(thos_batches),
+            )
+        ) as progress_bar:
+            sbc_outputs = Parallel(n_jobs=num_workers)(
+                delayed(sbc_on_batch)(
+                    thos_batch, xos_batch, posterior, num_posterior_samples
+                )
+                for thos_batch, xos_batch in zip(thos_batches, xos_batches)
+            )
+    else:
+        pbar = tqdm(
+            total=num_sbc_samples,
+            disable=not show_progress_bar,
+            desc=f"Running {num_sbc_samples} sbc samples.",
+        )
+
+        with pbar:
+            sbc_outputs = []
+            for thos_batch, xos_batch in zip(thos_batches, xos_batches):
+                sbc_outputs.append(
+                    sbc_on_batch(
+                        thos_batch,
+                        xos_batch,
+                        posterior,
+                        num_posterior_samples,
+                    )
+                )
+                pbar.update(sbc_batch_size)
+
+    # Aggregate results.
+    ranks = []
+    dap_samples = []
+    log_probs = []
+    for out in sbc_outputs:
+        ranks.append(out[0])
+        log_probs.append(out[1])
+        dap_samples.append(out[2])
+
+    ranks = torch.cat(ranks)
+    dap_samples = torch.cat(dap_samples)
+    log_probs = torch.cat(log_probs)
+
+    return ranks, log_probs, dap_samples
+
+
+def sbc_on_batch(
+    thos: Tensor, xos: Tensor, posterior: NeuralPosterior, L: int
+) -> Tuple[Tensor, Tensor, Tensor]:
+    """Return SBC results for a batch of SBC parameters and data from prior.
+
+    Args:
+        thos: ground truth parameters.
+        xos: corresponding observations.
+        posterior: sbi posterior.
+        L: number of samples to draw from the posterior in each sbc run.
+
+    Returns
+        ranks: ranks of true parameters vs. posterior samples under the specified RV,
+            for each posterior dimension.
+        log_prob_thos: log prob of true parameters under the approximate posterior.
+            Note that this is interpretable only for normalized log probs, i.e., when
+            using (S)NPE.
+        dap_samples: samples from the data averaged posterior for the current batch,
+            i.e., a single sample from each approximate posterior.
+    """
+
+    log_prob_thos = torch.zeros(thos.shape[0])
+    dap_samples = torch.zeros_like(thos)
+    ranks = torch.zeros_like(thos)
+
+    for idx, (tho, xo) in enumerate(zip(thos, xos)):
+        # Log prob of true params under posterior.
+        # NOTE: log_prob is implemented for all posteriors, but is unnormalized for all
+        # likelihood-based posteriors.
+        log_prob_thos[idx] = posterior.log_prob(tho, x=xo)
```

Comment:
this will trigger warnings for SNLE and SNRE. I suggest to use `.potential()` instead.

Reply:
this is intended. but true, at the moment, this warning pops up at every iteration. I will issue one warning manually instead.

Reply:
Why is this intended? SBC works with unnormalized log-probs, no? So what's the point of raising a warning?

Reply:
true, SBC itself does work with unnormalised, but calculating the log probs of true parameters under the posterior does not.